integration-tests/src/test/java/com/linkedin/kafka/clients/producer/LiKafkaInstrumentedProducerIntegrationTest.java

```diff
@@ -196,6 +198,70 @@ public void uncaughtException(Thread t, Throwable e) {
     Assert.assertTrue(root instanceof RecordTooLargeException, root.getMessage());
   }
 
+  @Test
+  public void testProducerFlushAfterClose() throws Exception {
```

Comment:
Have you tested what happens when the flush(timeout) is called when the close() is in progress and has not completed?

What does it return if timeout is less than the time it takes the close() to complete V/s timeout is more than the time it takes the close() to complete? In second case, will it return immediately after close() is completed or wait for the timeout

Reply:
It is overkill to have unit tests for all these scenarios given the fact that we might need to put extra wait time with different timeout and synchronization between threads.  we try to avoid "wait for timeout" in the unit tests as much as possible.  These two unit tests have covered the two new features:
1) allow timeoutexception to be propagated to users
2) allow flush to be called after close.

if you want to understand the behavior of flush when doing concurrent close,  we can write some local tests, and let you know the behavior of the following case:
1) a) close with timeout e.g., 10 seconds  in thread 1, b) flush with timeout e.g, 5 seconds in thread 2, see if flush hits a timeout.
2) a) close with timeout e.g. 5 seconds in thread 1, b) flush with timeout e.g., 10 seconds in thread 1, measure the time when flush finished.

Reply:
local tests will work. Thanks. I will wait to hear back about the behavior.

Reply:
The behavior is as follows:
1) case 1 -> flush time out
2) case 2 -> flush completed without a timeout.

example test code:

case 1

// close 
new Thread(() -> producer.close(Duration.ofSeconds(3))).start();
// wait close to in progress. 
Thread.sleep(1000);
producer.flush(1, TimeUnit.SECONDS);  ==> this timed out as expected.

case 2

// close 
new Thread(() -> producer.close(Duration.ofSeconds(3))).start();
// wait close to in progress. 
Thread.sleep(1000);
producer.flush(4, TimeUnit.SECONDS);  ==> this completes sucessfully without timeout.